torch/onnx/verification.py

```diff
@@ -0,0 +1,271 @@
+    )
+    # suppress ort warnings.
+    # 0:Verbose, 1:Info, 2:Warning. 3:Error, 4:Fatal. Default is 2.
+    so.log_severity_level = 3
+    ort_sess = onnxruntime.InferenceSession(
+        model_f if isinstance(model_f, str) else model_f.getvalue(),
+        so,
+        providers=ort_providers,
+    )
+    return ort_sess
+
+
+def _inline_flatten_list(inputs, res_list):
+    for i in inputs:
+        res_list.append(i) if not isinstance(
+            i, (list, tuple)
+        ) else _inline_flatten_list(i, res_list)
+    return res_list
+
+
+def _unpack_to_numpy(values):
+    value_unpacked = []
+    for value in values:
+        value_unpacked.extend(unpack_quantized_tensor(value))
+    return [_to_numpy(v) for v in value_unpacked]
+
+
+def _run_ort(ort_sess, inputs):
+    kw_inputs = {}
+    if inputs and isinstance(inputs[-1], dict):
+        kw_inputs = inputs[-1]
+        inputs = inputs[:-1]
+    inputs = _unpack_to_numpy(_flatten_tuples(inputs))
+    ort_inputs = {}
+    for input_name, input in kw_inputs.items():
+        ort_inputs[input_name] = _to_numpy(input)
+    inputs = _to_numpy(inputs)
+    ort_sess_inputs = ort_sess.get_inputs()
+    for i, input in enumerate(inputs):
+        if i == len(ort_sess_inputs) or ort_sess_inputs[i].name in ort_inputs:
+            raise ValueError(
+                f"got too many positional inputs. inputs: {inputs}. kw_inputs: {kw_inputs}"
+            )
+        ort_inputs[ort_sess_inputs[i].name] = input
+    ort_outs = ort_sess.run(None, ort_inputs)
+    return _inline_flatten_list(ort_outs, [])
+
+
+def _ort_compare_with_pytorch(ort_outs, output, rtol, atol):
+    output, _ = torch.jit._flatten(output)
+    outputs = _unpack_to_numpy(output)
+
+    # compare onnxruntime and PyTorch results
+    assert len(outputs) == len(ort_outs), "number of outputs differ"
+
+    # compare onnxruntime and PyTorch results
+    [
+        np.testing.assert_allclose(out, ort_out, rtol=rtol, atol=atol)
+        for out, ort_out in zip(outputs, ort_outs)
+    ]
+
+
+def verify(
+    model: Union[torch.nn.Module, torch.jit.ScriptModule],
+    input=None,
+    batch_size=2,
+    rtol=0.001,
+    atol=1e-7,
+    do_constant_folding=True,
+    dynamic_axes=None,
+    test_with_inputs=None,
+    input_names=None,
+    output_names=None,
+    fixed_batch_size=False,
+    dict_check=True,
+    training=None,
+    remained_onnx_input_idx=None,
+    flatten=True,
+    verbose=False,
+    ort_providers=_ORT_PROVIDERS,
+    ort_optim_on=True,
+    use_external_data=False,
+    opset_version: Optional[int] = None,
+    keep_initializers_as_inputs: bool = True,
+):
+    """Verify model export to ONNX with ONNXRuntime."""
+    if training is not None and training == torch.onnx.TrainingMode.TRAINING:
+        model.train()
+    elif training is None or training == torch.onnx.TrainingMode.EVAL:
+        model.eval()
+    if input is None:
+        input = torch.randn(batch_size, 3, 224, 224, requires_grad=True)
```

Comment:
Is the assumption on the input size a reasonable / general enough one?

Reply:
bc legacy code for caffe2 vision related model tests. I'll comment.

Reply:
removed in follow up